sympy/assumptions/cnf.py

```diff
@@ -0,0 +1,373 @@
+from itertools import combinations, product
+
+from sympy import S
+from sympy.logic.boolalg import BooleanFunction, Or, And, Not
+
+
+# Early encoding for literals
+literals_store = dict()
+
+
+class Literal(object):
+    """
+    The smallest element of a CNF object
+    """
+    literals_count = 0
+
+    def __new__(cls, lit, is_not=False):
+        if type(lit).__name__ == 'Not':
+            lit = lit.args[0]
+            is_not = True
+        k = literals_store.get(lit, None)
+        if k is not None:
+            code = k
+        elif lit in (S.false, False):
+            code = 0
+        else:
+            cls.literals_count += 1
+            code = cls.literals_count
+        obj = super(Literal, cls).__new__(cls)
+        obj.code = code
+        obj.lit = lit
+        literals_store[lit] = obj.code
+        obj.is_Not = is_not
+        return obj
+
+    @property
+    def arg(self):
+        sign = -1 if self.is_Not else 1
+        return sign*self.code
+
+    def __invert__(self):
+        is_not = not self.is_Not
+        return Literal(self.lit, is_not=is_not)
```

Comment:
Instead of storing is_not on the object, it would be simpler (and probably faster) to use negative integers for negations.